File: flutter/lib/api/internal/nudge/nudge_parser.dart
import 'package:flutter/material.dart';

import '../campaign/json_util.dart';
import 'nudge_config.dart';
import 'nudge_content.dart';

/// Decodes a nudge `templateConfig` (`{ container, layout }`) into the pure
/// [NudgeConfig] model.
///
/// This is the single place that knows the wire format. The model
/// (`nudge_content.dart`) and the renderer (`nudge_view.dart`) stay decoupled
/// from JSON, so the wire shape can evolve here without rippling outward (SRP).
///
/// Node decoding is dispatched through a registry ([_nodeParsers]) keyed by the
/// native widget type. Supporting a new nudge widget is an open/closed change:
/// add one entry to the map; nothing else in the parser is touched.
class NudgeParser {
  const NudgeParser();

  /// Returns `null` when the content tree is missing — such a campaign has
  /// nothing to show, so it is dropped rather than presented empty.
  NudgeConfig? parse(Map<String, dynamic> templateConfig) {
    final layout = optMap(templateConfig, 'layout');
    if (layout == null || layout.isEmpty) return null;

    return NudgeConfig(
      surface: _surface(optMap(templateConfig, 'container')),
      content: _column(layout),
    );
  }

  // ─── surface ───────────────────────────────────────────────────────────────

  NudgeSurface _surface(Map<String, dynamic>? json) {
    final map = json ?? const <String, dynamic>{};
    return NudgeSurface(
      displayType: optString(map, 'displayType', 'bottom_sheet') == 'dialog'
          ? NudgeDisplayType.dialog
          : NudgeDisplayType.bottomSheet,
      backgroundColor: _color(optString(map, 'backgroundColor')),
      cornerRadius: optDouble(map, 'cornerRadius', 18),
      padding: optDouble(map, 'padding', 20),
      backdropDismissible: optBool(map, 'backdropDismissible', true),
      showCloseButton: optBool(map, 'showCloseButton', false),
      showHandle: optBool(map, 'showHandle', true),
      draggable: optBool(map, 'draggable', true),
      // Stored as a 0…100 percentage; normalise to a 0…1 fraction.
      widthFraction: (optDouble(map, 'widthPct', 86) / 100).clamp(0.3, 1.0),
    );
  }

  // ─── column + children ───────────────────────────────────────────────────────

  NudgeColumn _column(Map<String, dynamic> layout) {
    final props = optMap(layout, 'props') ?? const <String, dynamic>{};
    return NudgeColumn(
      crossAxisAlignment: _crossAxis(optString(props, 'crossAxisAlignment', 'start')),
      mainAxisAlignment: _mainAxis(optString(props, 'mainAxisAlignment', 'start')),
      spacing: optDouble(props, 'spacing', 0),
      children: _childNodes(layout).map(_node).whereType<NudgeNode>().toList(),
    );
  }

  /// Reads a node's children whether they arrive as a bare list
  /// (`children: [...]`) or a child-group map (`children: { children: [...] }` /
  /// `childGroups: { children: [...] }`) — agnostic to which form was emitted.
  List<Map<String, dynamic>> _childNodes(Map<String, dynamic> node) {
    Object? raw = node['children'] ?? node['childGroups'];
    if (raw is Map) raw = raw['children'];
    if (raw is! List) return const [];
    return raw.whereType<Map>().map((m) => m.cast<String, dynamic>()).toList();
  }

  NudgeNode? _node(Map<String, dynamic> node) {
    final parse = _nodeParsers[optString(node, 'type')];
    if (parse == null) return null;
    final box = _box(optMap(node, 'containerProps'));
    return parse(box, optMap(node, 'props') ?? const <String, dynamic>{});
  }

  // ─── node parser registry (Strategy) ─────────────────────────────────────────

  static final Map<String, NudgeNode Function(NudgeBox, Map<String, dynamic>)>
      _nodeParsers = {
    'digia/text': _text,
    'digia/image': _image,
    'digia/button': _button,
    'fw/sized_box': _gap,
    'digia/styledHorizontalDivider': _divider,
    'digia/lottie': _lottie,
  };

  static NudgeNode _text(NudgeBox box, Map<String, dynamic> props) {
    final style = optMap(props, 'textStyle') ?? const {};
    final font = optMap(optMap(style, 'fontToken') ?? const {}, 'font') ?? const {};
    return NudgeText(
      box,
      text: optString(props, 'text'),
      fontSize: optDouble(font, 'size', 16),
      weight: _fontWeight(optString(font, 'weight', '400')),
      color: _color(optString(style, 'textColor')) ?? const Color(0xFF111111),
      align: _textAlign(optString(props, 'alignment', 'left')),
    );
  }

  static NudgeNode _image(NudgeBox box, Map<String, dynamic> props) => NudgeImage(
        box,
        url: optString(optMap(props, 'src') ?? const {}, 'imageSrc'),
        fit: _boxFit(optString(props, 'fit', 'cover')),
      );

  static NudgeNode _button(NudgeBox box, Map<String, dynamic> props) {
    final text = optMap(props, 'text') ?? const {};
    final textStyle = optMap(text, 'textStyle') ?? const {};
    return NudgeButton(
      box,
      label: optString(text, 'text', 'Button'),
      background: _color(optString(optMap(props, 'defaultStyle') ?? const {}, 'backgroundColor')) ??
          const Color(0xFF4945FF),
      textColor: _color(optString(textStyle, 'textColor')) ?? const Color(0xFFFFFFFF),
      radius: optDouble(optMap(props, 'shape') ?? const {}, 'borderRadius', 8),
      onClick: optMap(props, 'onClick'),
    );
  }

  static NudgeNode _gap(NudgeBox box, Map<String, dynamic> props) =>
      NudgeGap(box, height: optDouble(props, 'height', 8));

  static NudgeNode _divider(NudgeBox box, Map<String, dynamic> props) => NudgeDivider(
        box,
        thickness: optDouble(props, 'thickness', 1),
        indent: optDouble(props, 'indent', 0),
        endIndent: optDouble(props, 'endIndent', 0),
        color: _color(optString(optMap(props, 'colorType') ?? const {}, 'color')) ??
            const Color(0xFFE0E0E0),
      );

  static NudgeNode _lottie(NudgeBox box, Map<String, dynamic> props) => NudgeLottie(
        box,
        url: optString(optMap(props, 'src') ?? const {}, 'lottiePath'),
        height: optDouble(props, 'height', 160),
        loop: optString(props, 'animationType', 'loop') != 'once',
        autoplay: optBool(props, 'animate', true),
      );

  // ─── common box ──────────────────────────────────────────────────────────────

  NudgeBox _box(Map<String, dynamic>? containerProps) {
    final cp = containerProps ?? const <String, dynamic>{};
    final style = optMap(cp, 'style') ?? const <String, dynamic>{};
    final border = optMap(style, 'border');

    final widthStr = optString(style, 'width');
    return NudgeBox(
      fillWidth: widthStr == '100%',
      fixedWidth: widthStr == '100%' ? null : double.tryParse(widthStr),
      fixedHeight: double.tryParse(optString(style, 'height')),
      background: _color(optString(style, 'bgColor', optString(style, 'backgroundColor'))),
      padding: optDouble(style, 'padding', 0),
      margin: optDouble(style, 'margin', 0),
      borderRadius: optDouble(style, 'borderRadius', 0),
      borderColor: border == null ? null : _color(optString(border, 'borderColor')),
      borderWidth: border == null ? 0 : optDouble(border, 'borderWidth', 0),
      selfAlign: _selfAlign(optString(cp, 'align')),
    );
  }
}

// ─── enum + colour mapping ─────────────────────────────────────────────────────

CrossAxisAlignment _crossAxis(String value) => switch (value) {
      'center' => CrossAxisAlignment.center,
      'end' => CrossAxisAlignment.end,
      _ => CrossAxisAlignment.start,
    };

MainAxisAlignment _mainAxis(String value) => switch (value) {
      'center' => MainAxisAlignment.center,
      'end' => MainAxisAlignment.end,
      'spaceBetween' => MainAxisAlignment.spaceBetween,
      'spaceAround' => MainAxisAlignment.spaceAround,
      'spaceEvenly' => MainAxisAlignment.spaceEvenly,
      _ => MainAxisAlignment.start,
    };

TextAlign _textAlign(String value) => switch (value) {
      'center' => TextAlign.center,
      'right' || 'end' => TextAlign.right,
      _ => TextAlign.left,
    };

BoxFit _boxFit(String value) => switch (value) {
      'contain' => BoxFit.contain,
      'fill' => BoxFit.fill,
      _ => BoxFit.cover,
    };

FontWeight _fontWeight(String value) => switch (value) {
      '500' => FontWeight.w500,
      '600' => FontWeight.w600,
      '700' => FontWeight.w700,
      _ => FontWeight.w400,
    };

NudgeSelfAlign? _selfAlign(String value) => switch (value) {
      'start' => NudgeSelfAlign.start,
      'center' => NudgeSelfAlign.center,
      'end' => NudgeSelfAlign.end,
      _ => null,
    };

/// Parses `#RGB` / `#RRGGBB` / `#AARRGGBB`; empty/invalid → null.
Color? _color(String hex) {
  var value = hex.trim();
  if (value.isEmpty) return null;
  if (value.startsWith('#')) value = value.substring(1);
  if (value.length == 3) value = value.split('').map((c) => '$c$c').join();
  if (value.length == 6) value = 'FF$value';
  if (value.length != 8) return null;
  final intValue = int.tryParse(value, radix: 16);
  return intValue == null ? null : Color(intValue);
}

File: android/digia-engage/src/main/java/com/digia/engage/internal/model/NudgeParser.kt
package com.digia.engage.internal.model

import org.json.JSONArray
import org.json.JSONObject

/**
 * Decodes a nudge `templateConfig` ({ container, layout }) into [NudgeConfig].
 * Direct port of Flutter nudge_parser.dart — single place that knows the wire format.
 * No Android/Compose imports; pure data construction only.
 */
internal class NudgeParser {

    fun parse(templateConfig: JSONObject): NudgeConfig? {
        val layout = templateConfig.optJSONObject("layout") ?: return null
        return NudgeConfig(
            surface = parseSurface(templateConfig.optJSONObject("container")),
            content = parseColumn(layout),
        )
    }

    // ── surface ───────────────────────────────────────────────────────────────

    private fun parseSurface(json: JSONObject?): NudgeSurface {
        val j = json ?: JSONObject()
        val widthPct = j.optDouble("widthPct", 86.0)
        return NudgeSurface(
            displayType = NudgeDisplayType.fromString(j.optString("displayType")),
            backgroundColor = j.optString("backgroundColor").ifBlank { null },
            barrierColor = j.optString("barrierColor").ifBlank { null },
            cornerRadius = j.optDouble("cornerRadius", 18.0).toFloat(),
            padding = j.optDouble("padding", 20.0).toFloat(),
            backdropDismissible = j.optBoolean("backdropDismissible", true),
            showCloseButton = j.optBoolean("showCloseButton", false),
            showHandle = j.optBoolean("showHandle", true),
            draggable = j.optBoolean("draggable", true),
            widthFraction = (widthPct / 100.0).toFloat().coerceIn(0.3f, 1.0f),
        )
    }

    // ── column ────────────────────────────────────────────────────────────────

    private fun parseColumn(layout: JSONObject): NudgeColumn {
        val props = layout.optJSONObject("props") ?: JSONObject()
        return NudgeColumn(
            crossAxisAlignment = props.optString("crossAxisAlignment", "start"),
            mainAxisAlignment = props.optString("mainAxisAlignment", "start"),
            spacing = props.optDouble("spacing", 0.0).toFloat(),
            children = childNodes(layout).mapNotNull { parseNode(it) },
        )
    }

    private fun childNodes(node: JSONObject): List<JSONObject> {
        val raw: Any? = node.opt("children") ?: node.opt("childGroups")
        val list: Any? = if (raw is JSONObject) raw.opt("children") else raw
        if (list !is JSONArray) return emptyList()
        return (0 until list.length()).mapNotNull { list.optJSONObject(it) }
    }

    private fun parseNode(node: JSONObject): NudgeNode? {
        val type = node.optString("type")
        val box = parseBox(node.optJSONObject("containerProps"))
        val props = node.optJSONObject("props") ?: JSONObject()
        return nodeParsers[type]?.invoke(box, props)
    }

    // ── node registry (Strategy pattern) ─────────────────────────────────────

    private val nodeParsers: Map<String, (NudgeBox, JSONObject) -> NudgeNode> = mapOf(
        "digia/text" to ::parseText,
        "digia/image" to ::parseImage,
        "digia/button" to ::parseButton,
        "fw/sized_box" to ::parseGap,
        "digia/styledHorizontalDivider" to ::parseDivider,
        "digia/lottie" to ::parseLottie,
        "digia/carousel" to ::parseCarousel,
        "digia/videoPlayer" to ::parseVideo,
    )

    private fun parseText(box: NudgeBox, props: JSONObject): NudgeNode {
        val style = props.optJSONObject("textStyle") ?: JSONObject()
        val font = style.optJSONObject("fontToken")?.optJSONObject("font") ?: JSONObject()
        return NudgeText(
            box,
            text = props.optString("text"),
            fontSize = font.optDouble("size", 16.0).toFloat(),
            fontWeight = font.optString("weight", "400").toIntOrNull() ?: 400,
            color = parseColor(style.optString("textColor")) ?: 0xFF111111.toInt(),
            align = when (props.optString("alignment", "left")) {
                "center" -> NudgeTextAlign.CENTER
                "right", "end" -> NudgeTextAlign.RIGHT
                else -> NudgeTextAlign.LEFT
            },
        )
    }

    private fun parseImage(box: NudgeBox, props: JSONObject): NudgeNode {
        val ar = props.optDouble("aspectRatio", 0.0).toFloat()
        return NudgeImage(
            if (ar > 0f) box.withoutFixedHeight() else box,
            url = props.optJSONObject("src")?.optString("imageSrc") ?: "",
            fit = props.optString("fit", "cover"),
            aspectRatio = ar,
        )
    }

    private fun parseButton(box: NudgeBox, props: JSONObject): NudgeNode {
        val text = props.optJSONObject("text") ?: JSONObject()
        val textStyle = text.optJSONObject("textStyle") ?: JSONObject()
        val font = textStyle.optJSONObject("fontToken")?.optJSONObject("font") ?: JSONObject()
        val defaultStyle = props.optJSONObject("defaultStyle") ?: JSONObject()
        return NudgeButton(
            box,
            label = text.optString("text", "Button"),
            variant = when (props.optString("variant", "fill")) {
                "outline" -> NudgeButtonVariant.OUTLINE
                "text" -> NudgeButtonVariant.TEXT
                "elevated" -> NudgeButtonVariant.ELEVATED
                else -> NudgeButtonVariant.FILL
            },
            fontSize = font.optDouble("size", 16.0).toFloat(),
            fontWeight = font.optString("weight", "600").toIntOrNull() ?: 600,
            background = parseColor(defaultStyle.optString("backgroundColor")) ?: 0xFF4945FF.toInt(),
            textColor = parseColor(textStyle.optString("textColor")) ?: 0xFFFFFFFF.toInt(),
            radius = props.optJSONObject("shape")?.optDouble("borderRadius", 8.0)?.toFloat() ?: 8f,
            isPrimary = props.optBoolean("isPrimary", false),
            actions = NudgeActionParser().parse(props.optJSONObject("onClick")),
        )
    }

    private fun parseGap(box: NudgeBox, props: JSONObject): NudgeNode =
        NudgeGap(box, height = props.optDouble("height", 8.0).toFloat())

    private fun parseDivider(box: NudgeBox, props: JSONObject): NudgeNode = NudgeDivider(
        box,
        thickness = props.optDouble("thickness", 1.0).toFloat(),
        indent = props.optDouble("indent", 0.0).toFloat(),
        endIndent = props.optDouble("endIndent", 0.0).toFloat(),
        color = parseColor(props.optJSONObject("colorType")?.optString("color")) ?: 0xFFE0E0E0.toInt(),
    )

    private fun parseLottie(box: NudgeBox, props: JSONObject): NudgeNode = NudgeLottie(
        box,
        url = props.optJSONObject("src")?.optString("lottiePath") ?: "",
        height = props.optDouble("height", 160.0).toFloat(),
        loop = props.optString("animationType", "loop") != "once",
        autoplay = props.optBoolean("animate", true),
    )

    private fun parseCarousel(box: NudgeBox, props: JSONObject): NudgeNode {
        val imgs = mutableListOf<String>()
        props.optJSONArray("images")?.let { arr ->
            for (i in 0 until arr.length()) {
                val s = arr.optString(i)
                if (s.isNotEmpty()) imgs += s
            }
        }
        return NudgeCarousel(
            box,
            images = imgs,
            height = props.optDouble("height", 180.0).toFloat(),
            autoPlay = props.optBoolean("autoPlay", true),
            autoPlayInterval = props.optInt("autoPlayInterval", 3000),
            loop = props.optBoolean("infiniteScroll", true),
            showIndicator = props.optBoolean("showIndicator", true),
        )
    }

    private fun parseVideo(box: NudgeBox, props: JSONObject): NudgeNode = NudgeVideo(
        box,
        url = props.optString("url"),
        height = props.optDouble("height", 200.0).toFloat(),
        autoplay = props.optBoolean("autoPlay", false),
        loop = props.optBoolean("looping", false),
        showControls = props.optBoolean("showControls", true),
        muted = props.optBoolean("muted", false),
    )

    // ── box ───────────────────────────────────────────────────────────────────

    private fun parseBox(cp: JSONObject?): NudgeBox {
        val style = cp?.optJSONObject("style") ?: JSONObject()
        val border = style.optJSONObject("border")
        val widthStr = style.optString("width", "")
        return NudgeBox(
            fillWidth = widthStr == "100%",
            fixedWidth = if (widthStr == "100%") null else widthStr.toFloatOrNull(),
            fixedHeight = style.optString("height", "").toFloatOrNull(),
            background = parseColor(
                style.optString("bgColor", style.optString("backgroundColor", ""))
            ),
            paddingLeft = edgeSide(style.opt("padding"), "left"),
            paddingTop = edgeSide(style.opt("padding"), "top"),
            paddingRight = edgeSide(style.opt("padding"), "right"),
            paddingBottom = edgeSide(style.opt("padding"), "bottom"),
            marginLeft = edgeSide(style.opt("margin"), "left"),
            marginTop = edgeSide(style.opt("margin"), "top"),
            marginRight = edgeSide(style.opt("margin"), "right"),
            marginBottom = edgeSide(style.opt("margin"), "bottom"),
            borderRadius = style.optDouble("borderRadius", 0.0).toFloat(),
            borderColor = border?.let { parseColor(it.optString("borderColor")) },
            borderWidth = border?.optDouble("borderWidth", 0.0)?.toFloat() ?: 0f,
            selfAlign = NudgeSelfAlign.from(cp?.optString("align")),
        )
    }

    private fun edgeSide(raw: Any?, side: String): Float = when (raw) {
        is Number -> raw.toFloat()
        is JSONObject -> raw.optDouble(side, 0.0).toFloat()
        else -> 0f
    }
}

// ── Color helper ──────────────────────────────────────────────────────────────

/** Parses `#RGB` / `#RRGGBB` / `#AARRGGBB` hex strings; empty/invalid → null. */
internal fun parseColor(hex: String?): Int? {
    var v = hex?.trim() ?: return null
    if (v.isEmpty()) return null
    if (v.startsWith('#')) v = v.substring(1)
    if (v.length == 3) v = v.map { "$it$it" }.joinToString("")
    if (v.length == 6) v = "FF$v"
    if (v.length != 8) return null
    return v.toLongOrNull(16)?.toInt()
}
